documentation/design/proposals/global-console-phase1.adoc

```diff
@@ -0,0 +1,159 @@
+
+===== Recommendation
+
+I'm still thinking oauth-proxy.   I'd need to contant the Oauth-proxy team let them know of the use, and sound out the possiblity of an enhancement in future.  I would like the HTTPD sidecar to be as short term as possible).
+
+==== Operator integration
+
+The operator will be responsibilty for the installation of the global console and the creation of the resources it needs.  Its behaviour needs to be different depending on whether Openshift or a kubernetes platform is in use.
+
+===== Console CRD
+
+There will be a Console CRD.  When an instance of this CRD exists, the operator will trigger the installation of the global console.   It is antipicated that a cluster will have at most one instance of the console deployed.  What should happen if a second instance of the CRD is created? (Probably just log a warning and ignore it).
+
+On Openshift an instance of the Console CRD will be created automatically, thus bootstraping the install of the console on that platform.  On kubernetes, the user will create the CR by hand to install the console.
+
+If the CRe is removed, the Console will be undeployed by the operator.  On Openshift it won't be recreated automatically (might need state.  where would we keep it??).
+
+====== CRD Definition
+
+The CRD will encapsulate the following:
+
+- oauthDiscoverURL - a URL referencing an OAuth discovery metadat https://tools.ietf.org/id/draft-ietf-oauth-discovery-08.html.  Discovery metadata is becoming increasing common, but we'd allow the url to be a data-url so a user could construct their own discover document if need be.
+- oauthSecretRef - a path referencing either kubernetes secret instance or OpenShift openauthclient instance
+- scope - oauth scope
+
+
+On OpenShift `oauthDiscoverURL` would be optional.  The platform default https://openshift.default.svc/.well-known/oauth-authorization-server  would be used if not specified (see https://docs.openshift.com/container-platform/3.11/architecture/additional_concepts/authentication.html#oauth-server-metadata).  On kubernetes it would be mandatory.
+
+On OpenShift `oauthSecretRef` will reference an oauthclient object (cluster wide).  On kubernetes, it will reference a namespaced secret.  The operator will use this reference to retrieve the OAuth2 client id and client secret.
+
+===== Resources controlled by the Operator
+
+The Operator will use the information in the Console CR to create and keep synchronised the following resources.
+
+- oauthclient (OpenShift)
+- route (OpenShift only)
+- service (with serving-cert-secret-name on OpenShift).
+- serviceaccount
+- configmap instance 
+- deployment 
+
+On OpenShift, for the oauthclient, if the `oauthSecretRef` references an oauthclient instance that does not exist, the operator will create it. If it already exists, it will simply use it.  The Operator will ensure that the oauthclient's redirect urls are synchronized with the routes that exist to the service at that moment.
```

Comment:
Implementation note: this can probably share the existing oauth client used by the per-addressspace console? This is created by the operator today

Reply:
Yes, that's my thinking.   In fact, I think they *have* to share the oauthclient otherwise there won't be a seamless SSO experience. The oauthclient will need to be updated each time an addressspace comes or goes with the redirect url of the address space's console route.

When the user clicks through from global console to an addressspace console, there will be a OAuth2 dosey doe with the authentication server but the user won't need to authenticate.